[Synthetics] Monitor Add/Edit - Fix form error states

awahab07 opened the pull request:
Related to #154855

~Note: Should be merged post-GA, considering the magnitude of change.~

## Summary

- Connects form validation states with react-hoom-form's state (which was the reason of the reported bug that the error state was only reflected in euiForm and not the react-hook-form)
- Changes `reValidateMode` from `onChange` to `onSubmit` and manually invalidates the form when fields change so that dependent validations can be executed reactively e.g. fixes the following state:
  <img width="703" alt="Screenshot 2023-05-04 at 00 10 03" src="https://user-images.githubusercontent.com/2748376/236061192-a9417915-1eca-4cff-b871-b680955ec15a.png">
- Extracts validation and dependencies logic into a hook to simplify state management
- Fixes the following error where "Response body max bytes" field appear despite "Index response body" flag being unchecked.
  <img width="1205" alt="Screenshot 2023-05-05 at 19 35 27" src="https://user-images.githubusercontent.com/2748376/236527581-ff62550a-2679-431e-8571-da02a89c0a32.png">
- Fixes a case where updated values from for nested fields (such as "Response body contains JSON") weren't being passed to monitor inspect component ("Inspect configuration" flyout on monitor Add/Edit).
  
### Testing
- Since the core logic of the Monitor Add/Edit has been changed, the Add/Edit functionality should be thoroughly tested.
- Create at least one monitor of each type and ensure input validation works.
- Update above monitors and ensure everything works.
- Try to create a monitor with duplicate name and make sure the validation error is shown in time.
- Also test Getting Started page which appears while creating the first ever monitor in a Kibana space.

elasticmachine commented:
Pinging @elastic/uptime (Team:uptime)

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

dominiqueclarke reviewed (commented):
Noting that we are waiting to merge till for 8.9.0

awahab07 commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/fields/index_response_body_field.tsx:
```diff
@@ -25,15 +25,15 @@ export const ResponseBodyIndexField = ({
     if (checked) {
-      setPolicy(policy);
-      onChange(policy);
+      const defaultOrSelected =
```
Yes, it had a glitch.

awahab07 commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/field_config.tsx:
```diff
@@ -239,18 +239,15 @@ export const FIELD = (readOnly?: boolean): FieldMap => ({
-            });
+            setValue(ConfigKey.NAME, event.target.value);
           }
+          await trigger([ConfigKey.URLS, ConfigKey.NAME]);
```
With `setValue`, the form wasn't reflecting the most recent error states of dependent fields, especially when we also have to consider custom hook validations. This approach will leave us with better control over manually invalidating form whenever required, in custom/composite fields.

dominiqueclarke reviewed (approved):
LGTM!

awahab07 committed:
Make form fields validation states reactive.

awahab07 committed:
- Set *touched* to fields onChange so that errors can appear even before form submit.
- Serialize form data (MonitorFields) before passing to "Inspect configuration".

awahab07 committed:
Remove redundant useFormWrapped hook.

awahab07 committed:
Adjusted Getting Started form to comply with updated `useFormWrapped`.

shahzad31 reviewed (approved):
Smoke test LGTM !!

justinkambic commented:
#### Testing
- [x] Since the core logic of the Monitor Add/Edit has been changed, the Add/Edit functionality should be thoroughly tested.
- [x] Create at least one monitor of each type and ensure input validation works.
- [x] Update above monitors and ensure everything works.
- [x] Try to create a monitor with duplicate name and make sure the validation error is shown in time.
- [x] Also test Getting Started page which appears while creating the first ever monitor in a Kibana space.
